design/src/rfcs/rfc0017_logging_sensitive.md

```diff
@@ -0,0 +1,376 @@
+
+## Proposal
+
+This proposal serves to honor the sensitivity specification via code generation of a logging `Layer`s which is aware of the sensitivity, together with a developer contract disallowing logging potentially sensitive data in the runtime crates. An internal and external guideline should be provided in addition to the `Layer`s.
+
+All data known to be sensitive should be replaced with `"_redacted_"` when logged. Implementation wise this means that [tracing::Event](https://docs.rs/tracing/latest/tracing/struct.Event.html)s and [tracing::Span](https://docs.rs/tracing/latest/tracing/struct.Span.html)s of the form `debug!(field = "sensitive data")` and `span!(..., field = "sensitive data")` must become `debug!(field = "_redacted_")` and `span!(..., field = "_redacted_")`.
+
+### Debug Logging
+
+Developers might want to observe sensitive data for debugging purposes. It should be possible to opt-out of the redactions by enabling a feature flag `debug-logging` (which is disabled by default).
+
+To prevent excessive branches such as
+
+```rust
+if cfg!(feature = "debug-logging") {
+    debug!(%data, "logging here");
+} else {
+    debug!(data = "_redacted_", "logging here");
+}
+```
+
+the following wrapper should be provided
+
+```rust
+pub struct Sensitive<T>(T);
+
+impl<T> Debug for Sensitive<T>
+where
+    T: Debug
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
+        if cfg!(feature = "debug-logging") {
+            self.0.fmt(f)
+        } else {
+            "_redacted_".fmt(f)
+        }
+    }
+}
+
+impl<T> Display for Sensitive<T>
+where
+    T: Display
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
+        if cfg!(feature = "debug-logging") {
+            self.0.fmt(f)
+        } else {
+            "_redacted_".fmt(f)
+        }
+    }
+}
+```
+
+In which case the branch above becomes
+
+```rust
+debug!(sensitive_data = %Sensitive(data));
+```
+
+### Code Generated Logging Layer
+
+Using the smithy model, for each operation, a logging `Layer` should be generated. Through the model, the code generation knows which fields are sensitive and which HTTP bindings exist, therefore the `Logging` layer can be careful crafted to avoid leaking sensitive data.
+
+This `Layer` should be applied to the [OperationHandler](https://github.com/awslabs/smithy-rs/blob/cd0563020abcde866a741fa123e3f2e18e1be1c9/rust-runtime/inlineable/src/server_operation_handler_trait.rs#L17-L21) directly after its construction in the generated `operation_registry.rs`. The `Layer` should preserve the associated types of the `OperationHandler` (`Response = Response<BoxBody>`, `Error = Infallible`) so cause minimal breakage.
+
+This is illustrated before as the introduction of `Logging{Operation}::new`.
+
+```rust
+let empty_operation = LoggingEmptyOperation::new(operation(registry.empty_operation));
```

Comment:
One of the problems with putting it here is that any layer applied on top of this will fail to included in the `span!`. What we really want is for the logging layer to **always** be applied just after the routing has been performed.

Reply:
I think this is something quite serious and should be addressed. I'm leaving a comment in the RFC and adding it to the checklist.